Setup new dotnet new android measurements using Mono or CoreCLR runtimes

matouskozak opened the pull request:
Switching MAUI-Android (and .NET Android) scenarios to use latest .NET SDK + dotnet10 feed, the steps are:
1. Install the latest .NET SDK: `.dotnet-install.sh --channel 10.0 --quality daily`
2. Using the latest feed ([dotnet10](https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet10/nuget/v3/index.json)), find the latest packs using `dotnet package search`.
3. Put the latest packs to a rollback file and install maui workload.

Adding new jobs:
- `dotnet new android` running CoreCLR.
- `dotnet new maui` (Android) running CoreCLR.
- `dotnet new maui-blazor` (Android) running CoreCLR.

Internal pipeline run: https://dev.azure.com/dnceng/internal/_build/results?buildId=2684265

TODO:
- [x] Remove added android feeds from `Nuget.config` (should be added by maestro)

Out of scope
- Fix flaky memory consumption measurements for .NET Android and .NET MAUI Android https://github.com/dotnet/performance/issues/4826

---

https://github.com/dotnet/performance/issues/4754

LoopedBard3 commented on src/scenarios/shared/precommands.py:
```diff
@@ -288,6 +288,37 @@ def uninstall_workload(self, workloadid: str):
         if self.has_workload and not self.readonly_dotnet:
             subprocess.run(["dotnet", "workload", "uninstall", workloadid])
 
+    def create_nuget_config(self, feeds: dict[str, str]={"nuget.org": "https://api.nuget.org/v3/index.json"}):
```
We will likely get workitems to pop up if we use nuget.org, would it instead be possible to instead use the one at the top level of the performance repo? If there are some packages that only occasionally update but they are not being found in our feeds, we can have them added.

Also, this does not appear to currently be called anywhere.

matouskozak replied:
removed

LoopedBard3 commented on scripts/run_performance_job.py:
```diff
@@ -541,6 +541,17 @@ def run_performance_job(args: RunPerformanceJobArgs):
+    if args.run_kind == "maui_scenarios_android":
+        if args.runtime_flavor == "mono":
+            runtime_type = "Mono"
+            configurations["CompilationMode"] = "ProfiledAOT"
```
We should make sure other similar android app testing matches this for Mono.

matouskozak replied:
I added a comment explaining which scenario is which. We run different configuration for dotnet/runtime android sample than the default configuration used in .NET Android / .NET MAUI Android samples.

caaavik-msft reviewed (dismissed):
Looks good to me. One question I have, are any of these changes (e.g. the _MSBuildArgs) being made because they are coupled to the runtime repository itself or would you want them set even if we decided to run against an older version of the runtime? 

As part of the recent changes we made to move the yaml from the runtime to the performance repository, I'd like to know if any of the things we moved over were better living in the runtime repository. If so, we can look at restructuring where these live to better couple related changes.

ivanpovazan commented on eng/performance/maui_scenarios_android.proj:
```diff
@@ -11,6 +11,10 @@
     <PreparePayloadWorkItemBaseDirectory Condition="'$(TargetsWindows)' != 'true'">$(CorrelationPayloadDirectory)$(PreparePayloadOutDirectoryName)/</PreparePayloadWorkItemBaseDirectory>
+
+    <_MSBuildArgs Condition="'$(RuntimeFlavor)' == 'mono'">/p:UseMonoRuntime=true</_MSBuildArgs>
+    <_MSBuildArgs Condition="'$(RuntimeFlavor)' == 'coreclr'">/p:UseMonoRuntime=false;/p:AndroidEnableMarshalMethods=false</_MSBuildArgs>
```
```suggestion
    <_MSBuildArgs Condition="'$(RuntimeFlavor)' == 'coreclr' and '$(PublishAot)' != 'true'">/p:UseMonoRuntime=false</_MSBuildArgs>
```
`AndroidEnableMarshalMethods ` is not needed anymore as of: https://github.com/dotnet/android/pull/9912

ivanpovazan commented on eng/performance/maui_scenarios_android.proj:
```diff
@@ -25,19 +29,19 @@
-      <PackageName>com.companyname.mauiandroiddefault</PackageName>
-    </MAUIAndroidScenario>
-    <MAUIAndroidScenario Include=".NET Android Default">
+    <MAUIAndroidScenario Include=".NET Android Default $(RunConfigsString)">
```
I think we should revisit the naming here and/or introduce new cases.
The `Default` is different between Mono, CoreCLR and NativeAOT, Mono's defaults to profiled AOT, while CoreCLR defaults to JIT only *at the moment* etc, which can lead to confusion when analysing the measurements.
I suggest we add new more descriptive scenarios (JIT, AOT, ..) conditioned on the runtime type.
Lastly, marking which one of the scenarios is the default one is still valueable.

matouskozak replied:
Added

matouskozak replied:
Actually, I'm thinking about removing the `$(RunConfigsString)` all together from the scenario name as per 
https://github.com/dotnet/runtime/pull/114117#discussion_r2023567496 and offline discussion with @LoopedBard3 . 

We can already see the configuration used in the job name 
<img width="611" alt="image" src="https://github.com/user-attachments/assets/59cd5c4a-2864-4c7d-9130-7f9d453dd2b0" />
+ everything we place in `configurations` https://github.com/dotnet/performance/blob/2659b9d72210c725f20e7a10499ff88d40bcc29f/scripts/run_performance_job.py#L550-L551 is going to flow to database so we can retrieve it for visualizations...

ivanpovazan commented on eng/performance/maui_scenarios_android.proj:
```diff
@@ -25,19 +29,19 @@
+      <ApkName>com.companyname.mauiandroiddefault-Signed</ApkName>
+      <PackageName>com.companyname.mauiandroiddefault</PackageName>
+    </MAUIAndroidScenario>
+    <MAUIAndroidScenario Include="Maui Blazor Android Default (TTFD) $(RunConfigsString)" Condition="'$(RuntimeFlavor)' == 'mono'">
```
Why do we only care about TTFD measurement here? Or why don't we measure both TTID and TTFD?

matouskozak replied:
I don't know. I just saw that it is passing the `--use-fully-drawn-time` so I added it to the scenario name to make it clear. I can probably add TTID as well if this is something we want.

matouskozak replied:
@LoopedBard3 do you remember why for blazor we are using `--use-fully-drawn-time` and not the default startup measurement config?

LoopedBard3 replied:
I don't recall anything specifically. Maybe TTID doesn't include some of the blazor startup. Regardless, I don't see any reason not to include both.

ivanpovazan commented on scripts/run_performance_job.py:
```diff
@@ -531,16 +531,30 @@ def run_performance_job(args: RunPerformanceJobArgs):
+    if args.run_kind == "maui_scenarios_android":
+        if args.runtime_flavor == "mono":
+            runtime_type = "Mono"
+            configurations["CompilationMode"] = "ProfiledAOT"
```
Can we use `CompilationMode` in the `MAUIAndroidScenario` naming above?

ivanpovazan commented on src/scenarios/mauiandroid/pre.py:
```diff
@@ -11,9 +11,14 @@
 precommands = PreCommands()
-install_versioned_maui(precommands)
+
+precommands.setup_workload_update_mode("manifests")
```
Should this be the default behavior? 
My understanding is that we have to set the update mode only if we want to use the nightly builds

matouskozak replied:
> My understanding is that we have to set the update mode only if we want to use the nightly builds

This is true. 

> Should this be the default behavior?

The change is intended to be the new default, and I wasn't planning for use to have 2 separate pipelines, one of released and another of nightly builds. These measurements should serve as a tool for internal monitoring and not public benchmarks. Do you see benefits in having both scenarios (released/nightly) running in the lab?

ivanpovazan replied:
If we checkout the repo and want to run the scenario locally with a specific release how would we do that?

matouskozak replied:
I see, I have not considered this as a user scenario when designing this change. 

We can specify local dotnet install by `. ./init.sh -dotnetdir <custom dotnet directory>` so I think it could be enough to pass an argument to the `pre.py` command to skip installing the additional workloads and rely on the user that the `<custom dotnet directory>` contains the specific release installed. Would that be sufficient for the use case you have in mind?

matouskozak replied:
Update on that, we actually already support this scenario. If you use:
1. `. ./init.sh -dotnetdir <custom dotnet directory>`
2. `pre.py <other_args> --readonly-dotnet --has-workload`

This should ensure that the test app is build using the locally installed dotnet that has the specific configuration of SDKs.

ivanpovazan commented:
@matouskozak 

There might be some issues with the approach of `dotnet workload config --update-mode manifests` especially around servicing releases, because the manifests will include a reference to a new servicing release (like atm `9.0.4`) which is not publicly available so the workload installation would fail with:
```
Workload installation failed: One or more errors occurred. (Version 9.0.4 of package microsoft.net.runtime.monoaotcompiler.task is not found in NuGet feeds ...
```
Will experiment a bit with this but I think we might need to the the `Option 2` from your proposal in: https://github.com/dotnet/performance/issues/4754

matouskozak commented:
Refactored the PR to address some issues and changed the maui installation to use rollback file instead. The PR contains changes to Nuget.config which should be done maestro so should be reverted before merging.

ivanpovazan commented on eng/performance/maui_scenarios_android.proj:
```diff
@@ -11,6 +11,10 @@
     <PreparePayloadWorkItemBaseDirectory Condition="'$(TargetsWindows)' != 'true'">$(CorrelationPayloadDirectory)$(PreparePayloadOutDirectoryName)/</PreparePayloadWorkItemBaseDirectory>
+
+    <_MSBuildArgs Condition="'$(RuntimeFlavor)' == 'mono'">/p:UseMonoRuntime=true</_MSBuildArgs>
+    <_MSBuildArgs Condition="'$(RuntimeFlavor)' == 'coreclr'">/p:UseMonoRuntime=false;</_MSBuildArgs>
```
```suggestion
    <_MSBuildArgs Condition="'$(RuntimeFlavor)' == 'coreclr'">/p:UseMonoRuntime=false</_MSBuildArgs>
```

Do we need to concat more than this property?

ivanpovazan commented on scripts/run_performance_job.py:
```diff
@@ -531,16 +531,25 @@ def run_performance_job(args: RunPerformanceJobArgs):
+    if args.run_kind == "maui_scenarios_android":
+        if not args.runtime_flavor in ("mono", "coreclr"):
+            raise Exception("Runtime flavor must be specified for maui_scenarios_android")
+        configurations["CodegenType"] = str(args.codegen_type)
```
Just for my understanding how is `CodegenType` different from `CompilationMode` used above?

ivanpovazan replied:
Ok, I was referring to lines:
https://github.com/dotnet/performance/pull/4770/files#diff-75e8f68d6caa7704e8c72407bcfc108e368c9b3ea7e71ca1532033c8905bad46R543
https://github.com/dotnet/performance/pull/4770/files#diff-75e8f68d6caa7704e8c72407bcfc108e368c9b3ea7e71ca1532033c8905bad46R537

matouskozak replied:
I think these are old version, `CompilationMode` is no longer being used there.

ivanpovazan commented on src/scenarios/shared/mauisharedpython.py:
```diff
@@ -93,4 +94,80 @@ def install_versioned_maui(precommands: PreCommands):
-    precommands.install_workload('maui', workload_install_args) 
+    precommands.install_workload('maui', workload_install_args)
+
+# TODO: the feed should be updated to the latest
```
Once we have darc setup to update NuGet feeds, this won't be needed anymore, correct?

matouskozak replied:
I wrote a new logic which extracts the latest feed from the repo's Nuget.config so we don't need this anymore.

ivanpovazan reviewed (approved):
LGTM, only left few minor notes.
Thank you!

LoopedBard3 commented on src/scenarios/shared/versionmanager.py:
```diff
@@ -34,4 +34,63 @@ def get_version_from_dll_powershell(dll_path: str):
+
+        version, commit = parse_version_output(result)
+        results[f"{sdk}_version"] = version
+        results[f"{sdk}_commit_hash"] = commit
```
Rather than adding a new string to look for in the Reporter.cs, can this instead be `PERFLAB_DATA_{sdk}_commit_hash`? The PERFLAB_DATA_ is not included in the saved additional data so it should show up the same, but then we don't need to add the additional search string `_commit_hash`.

LoopedBard3 reviewed (commented):
This looks really great! I am excited to have what seems to be a more robust solution for building the apps. I left one comment but am also wondering if this supports 9.0 and 8.0 mobile app runs. It seems that it currently sets up the runs to only run main.

matouskozak commented:
>  I left one comment but am also wondering if this supports 9.0 and 8.0 mobile app runs.

Yes, currently works just on main. I think it could be extended to other branches. We would just search a different dotnet feed I think and make it more parametrized. However, I don't see much value now in having measurements on other branches but we could extend it in the future if needed.

ivanpovazan reviewed (approved):
This looks neat, thank you!

The only thoughts left (not for this PR) are:
1. It seems like the rollback file could be provided by some other service rather than us manually creating it as part of performance runs 
2. Think about a way of doing performance runs on specific releases, versions i.e., answering a question: what is the performance of our preview.X release - and with that freeze/tag such measurements for reference

Other than that, really nice work!

caaavik-msft reviewed (dismissed):
LGTM, just the one minor comment on the yaml condition
